File: Num_Analysis/X4.py

def Gaussian_elim(A, b):
    """
    Gaussian elimination with partial pivoting
    
    Solves a system of linear equations Ax = b using Gaussian elimination with partial pivoting.

    @arg: A--Matrix representing the coefficients of the linear equations
    @arg: b--Vector representing the constants on the right-hand side of the equations
    """
    print(f'Input Matrix: {A} \nInput Vector: {b}')
    # Get the size of the matrix
    n = len(A)

    # Initialize a modified matrix M to hold augmented matrix [A | b]
    M = A

    # Append the elements of b to the corresponding rows of M to form the augmented matrix
    i = 0
    for x in M:
        x.append(b[i])
        i += 1

    for k in range(n): # start in left column move to right
        for i in range(k, n): # loop though elements in column checking for larger elements that the pivot
            if abs(M[i][k]) > abs(M[k][k]):
                M[k], M[i] = M[i], M[k]

        
        for j in range(k + 1, n): # Eliminate the lesser elements than the pivot
            q = float(M[j][k]) / M[k][k]
            for m in range(k, n + 1):
                M[j][m] -= q * M[k][m]

    # Back substitution
    x = [0 for i in range(n)] # This will be the solution

    x[n - 1] = float(M[n - 1][n]) / M[n - 1][n - 1] # solve for last unknown

    for i in range(n - 2, -1, -1): # solve for remaining unknowns
        z = 0
        for j in range(i + 1, n):
            z += float(M[i][j]) * x[j]
        x[i] = float(M[i][n] - z) / M[i][i]
    
    # Print the solution vector
    print(f'Solution: {x}')

# Book Example 3.1.4: Book solution: 1/9, 5/9, 0, 1/3
# Code solution [0.11111111111111108, 0.5555555555555556, -8.034508730839949e-17, 0.33333333333333337]
A = [[2, 5, 5, 3],[3, 3, 1, 3], [1, 4, 1, 2],[5,5,1,2]]
b = [4,3,3,4]

Gaussian_elim(A, b)

File: Num_Analysis/X4.jl

function Gaussian_elim(A, b)
    """
    Gaussian elimination with partial pivoting

    Solves a system of linear equations Ax = b using Gaussian elimination with partial pivoting.

    Arguments:
    - A: Matrix representing the coefficients of the linear equations
    - b: Vector representing the constants on the right-hand side of the equations
    """
    println("Input Matrix A size: ", size(A), "\nInput Vector b size: ", size(b))
    # Get the size of the matrix
    n = size(A, 1)

    # Initialize a modified matrix M to hold augmented matrix [A | b]
    M = [A b]

    println("Augmented Matrix M size: ", size(M))

    for k in 1:n # start in left column move to right
        pivot = M[k, k]
        for i in k+1:n # loop through elements in column checking for larger elements that the pivot
            if abs(M[i, k]) > abs(pivot)
                pivot = M[i, k]
                M[k, :], M[i, :] = M[i, :], M[k, :]  # Swap rows
            end
        end

        for j in k + 1:n # Eliminate the lesser elements than the pivot
            q = M[j, k] / pivot
            M[j, k:end] .-= q * M[k, k:end]
        end
    end

    # Back substitution
    x = zeros(n) # This will be the solution
    x[n] = M[n, end] / M[n, n] # solve for last unknown

    println("Solution vector x size: ", size(x))

    for i in n - 1:-1:1 # solve for remaining unknowns
        z = 0
        for j in i+1:n
            z += M[i, j] * x[j]
        end
        x[i] = (M[i, end] - z) / M[i, i]  # Corrected index for last column of M
    end

    # Print the solution vector
    println("Solution: ", x)
end

# Example usage
A = [2.0 5.0 5.0 3.0; 3.0 3.0 1.0 3.0; 1.0 4.0 1.0 2.0; 5.0 5.0 1.0 2.0]
b = [4.0, 3.0, 3.0, 4.0]  # Float64 values

# Book Example 3.1.4: Book solution: 1/9, 5/9, 0, 1/3
# Code Solution: [0.11111111111111108, 0.5555555555555556,
# -8.034508730839949e-17, 0.33333333333333337]

Gaussian_elim(A, b)

